mixer/pkg/runtime/dispatcher/session.go

```diff
@@ -333,7 +335,31 @@ func (s *session) waitForDispatched() {
 			if buf == nil {
 				buf = pool.GetBuffer()
 				// the first failure result's code becomes the result code for the output
+				// `buf` variable guards the first failure since it is set the first time
 				code = rpc.Code(st.Code)
+
+				// update the direct response matching the error status
+				if s.variety == tpb.TEMPLATE_VARIETY_CHECK {
+					if got, response := status.GetDirectHTTPResponse(st); got {
+						if s.checkResult.RouteDirective == nil {
+							s.checkResult.RouteDirective = &mixerpb.RouteDirective{}
+						}
+						directive := s.checkResult.RouteDirective
+						if response.Code != 0 {
+							directive.DirectResponseCode = uint32(response.Code)
+						} else {
+							directive.DirectResponseCode = uint32(gatewayruntime.HTTPStatusFromCode(codes.Code(response.Code)))
```

Comment:
Is this always 0 or there could be some other values?

Reply:
Thank you, this is indeed a bug. It should be using status RPC code by default, not 200.
Added a unit test to ensure all code is covered.